### PR TITLE
[ISSUE#4824] Support consumequeue, indexfile multiple directories storage  mode

### PR DESCRIPTION
**Pull request by shengminw**
[ISSUE#4824](https://github.com/apache/rocketmq/issues/4824)

In RIP-7 https://github.com/apache/rocketmq/wiki/RIP-7-Multiple-Directories-Storage-Support, CommitLog File has supported multi-directories storage mode.
In this case, the commitlog can be stored on multiple disks, but consumequeue/index is still stored on one disk. However, the size of consumequeue/index can reach tens of Gbs, or even 100 Gbs, and the space used can't be easily ignored.
The current commitLog multipath needs to be configured independently, as follows:
```java
storePathRootDir=/rmqstore1
storePathCommitLog=/rmqstore1/commitlog:/rmqstore2/commitlog
```
Therefore, I optimized the multi-path mode, and the improvements are as follows:
● Add multi-path mode for consumequeue.  
● Add multi-path mode for indexfile.
● Optimize the method to get the commitlog path. It is not necessary to specify storePathCommitLog separately, and generate the storePathCommitLog from storePathRootDir.
● Add a Timed thread to check path capacity. When select store path, it will  try to avoid writing to a nearly full disk.
● Loading  metadata in multi-path. By default, the config file is stored in the first path, and the config bak file is stored in the second path. For example: storePathRootDir = /rmqstore0:/rmqstore1:/rmqstore2, config dir = /rmqstore0/config, config bak dir = /rmqstore1/config. At startup, scan all paths and load the config file from the latest path of dataversion.（This ensures that when the path is changed, the latest version can be loaded）
● Loading checkpoint file and abort file in multi-path. The loading path of the checkpoint file and the abort file is the same as that of the config file.

## Chinese Version / 中文版本
在RIP-7 https://github.com/apache/rocketmq/wiki/RIP-7-Multiple-Directories-Storage-Support 当中，CommitLog文件已经支持了多文件路径的存储模式。
在这个情况下，commitLog可以被存储至多块磁盘当中，但是consumequeue和index依然存储于单盘上。然而consumequeue 和 index的文件大小也可能达到数十G或者上百G的大小。这样的存储空间并不能被忽略。
当前的commitLog的多路径需要被额外配置，如下所示：
```java
storePathRootDir=/rmqstore1
storePathCommitLog=/rmqstore1/commitlog:/rmqstore2/commitlog
```
因此，我对多路径模式做了一定的优化，改进主要如下：
● consumequeue的存储增加多路径模式。
● indexfile的存储增加多路径模式。
● 优化commitlog获取文件路径的方式。不必要单独设置storePathCommitLog，storePathCommitLog可以从storePathRootDir根目录下直接生成。
● 增加一个定时线程检查路径的容量。当选择写入路径的时候，尽量避免写入至快满的磁盘中。
● 多路径下的元数据加载。默认将config文件夹写入至第1个路径中，config文件夹的bak文件写入至第2个路径。例如：storePathRootDir = /rmqstore0:/rmqstore1:/rmqstore2, config dir = /rmqstore0/config, config bak dir = /rmqstore1/config。在启动的时候，扫描所有路径，选择dataversion最新的所在路径作为初始加载路径。（这样能确保路径被更改时，能够从最新的路径下加载）
● 加载checkpoint文件和abort文件。checkpoint文件和abort文件的加载路径与config路径保持一致。

### CODE REVIEW

**Comment by zongtanghu** [comment]
Can you fix the codes' conflicts firstly! @shengminw

**Comment by shengminw** [comment]
> Can you fix the codes' conflicts firstly! @shengminw

Sorry, this conflict was newly created when the develop branch was merged yesterday. I have fixed this conflict.

File: broker/src/main/java/org/apache/rocketmq/broker/filter/ConsumerFilterManager.java [hunk 1]
```diff
@@ -226,6 +226,16 @@ public String configFilePath() {
...
     }
 
+    @Override
+    public String configFileBakPath() {
+        if (this.brokerController != null) {
+            return BrokerPathConfigHelper.getConsumerFilterBakPath(
+                    this.brokerController.getMessageStoreConfig().getStorePathRootDir()
+            );
+        }
+        return BrokerPathConfigHelper.getConsumerFilterBakPath("./unit_test");
```
Comment (aaron-ai) [inline broker/src/main/java/org/apache/rocketmq/broker/filter/ConsumerFilterManager.java:236]: is "./unit_test" here compatible with windows platform?

File: broker/src/main/java/org/apache/rocketmq/broker/offset/ConsumerOrderInfoManager.java [hunk 2]
```diff
@@ -266,6 +266,15 @@ public String configFilePath() {
         }
     }
 
+    @Override
+    public String configFileBakPath() {
+        if (brokerController != null) {
+            return BrokerPathConfigHelper.getConsumerOrderInfoBakPath(this.brokerController.getMessageStoreConfig().getStorePathRootDir());
+        } else {
+            return BrokerPathConfigHelper.getConsumerOrderInfoBakPath("~");
```
Comment (aaron-ai) [inline broker/src/main/java/org/apache/rocketmq/broker/offset/ConsumerOrderInfoManager.java:274]: ditto

File: broker/src/test/java/org/apache/rocketmq/broker/BrokerControllerTest.java [hunk 3]
```diff
@@ -67,6 +71,65 @@ public void testBrokerRestart() throws Exception {
         brokerController.shutdown();
     }
 
+    @Test
+    public void testMultiPathStart() throws Exception {
+        MessageStoreConfig messageStoreConfig = new MessageStoreConfig();
+        messageStoreConfig.setStorePathRootDir(
+                "target/unit_test_store/a/store" + MixAll.MULTI_PATH_SPLITTER
```
Comment (aaron-ai) [inline broker/src/test/java/org/apache/rocketmq/broker/BrokerControllerTest.java:78]: ditto

File: broker/src/test/java/org/apache/rocketmq/broker/BrokerControllerTest.java [hunk 4]
```diff
@@ -67,6 +71,65 @@ public void testBrokerRestart() throws Exception {
...
+        brokerController.shutdown();
+        UtilAll.deleteFile(new File(new MessageStoreConfig().getStorePathRootDir()));
+    }
+
+    @Test
+    public void testMultiPathLoadConfig() {
+        MessageStoreConfig messageStoreConfig = new MessageStoreConfig();
+        messageStoreConfig.setStorePathRootDir(
+                "target/unit_test_store/a/store" + MixAll.MULTI_PATH_SPLITTER
+                        + "target/unit_test_store/b/store" + MixAll.MULTI_PATH_SPLITTER
```
Comment (aaron-ai) [inline broker/src/test/java/org/apache/rocketmq/broker/BrokerControllerTest.java:98]: ditto

**Comment by aaron-ai** [comment]
We intend to add CI pipelines for more platforms recently besides Linux, there are too many unit test could not pass in other platforms **especially Windows**.

It would be awesome if you could concern more about cross-platform issues.

By the way, if you don't have windows device for testing now, Github Action is a good choice. You can refer to https://github.com/aaron-ai/rocketmq/pull/1. I am pleased to provide support.

**Comment by shengminw** [comment]
@aaron-ai thanks for your suggestion, I will check the compatible with windows platform.

**Pull request closed without merging**